### PR TITLE
Structure_oM: Refactor bar reinforcement

### LINKED ISSUE #579: Structure_oM: Reinforcement functionality
#### Description:
BHoM currently "has" functionality for Reinforcements in concrete, but you currently can't change the minimum cover value, resulting in sections like this:
ConcreteWrong1
And T-sections can similarly not place its reinforcement within the "tie area"
ConcreteWrong2

So as I will need to use sections with reinforcement I'm planing to rectify this and expand to make it usable.

### The Plan:

- Move the `MinimumCover` variable from the section to the reinforcement object
- Fix the Create methods with regard to that
- new functional Create methods for things like T-sections
- Update the `Layout()` method which currently places the reinforcement "circles" as it is currently marked as `[NotImplemented]` (yet used)
- Create functionality to extract reinforcement positions from the section object (should not have to be done per toolkit for sanity reasons) (Might be a part of `Layout()` already

#### Secondary smaller Plan

- Calculate the section properties for cracked sections in stage II & III

And since it does not seem to be used in the rest of the BHoM at the moment but will be used after this, it would be nice to get various suggestions on how to restructure it while it doesn't disturb anything, so that we don't have to do it again in the foreseeable future.

@IsakNaslundBh @rwemay  @al-fisher @pawelbaran @JosefTaylor @peterjamesnugent

### LINKED ISSUE #755: Physical_oM: Add Rebar object
#### Description:

New object for a single Rebar to be added, after discussion with @pawelbaran and @LMarkowski 

Proposed object:
```
    public class Rebar : BHoMObject
    {
        /***************************************************/
        /**** Properties                                ****/
        /***************************************************/

        public ICurve CentreCurve { get; set; }

        public double Diameter { get; set; }

        public Steel Material { get; set; }

        /***************************************************/
    }
```
Related to https://github.com/BHoM/BHoM/issues/579

### PR DESCRIPTION
**Pull request by IsakNaslundBh**
### Issues addressed by this PR

Closes #579
Closes #755 

### Test files

https://burohappold.sharepoint.com/:f:/s/BHoM/EpXy3FR30txGs3Q3jTlTwOoBv9zjDu3k-0Z34O9CfSBIBQ?e=OO58dZ

### Changelog

- Complete refactor of bar Reinforcement objects making use of layouts in spacial_oM
- All old reinforcement classes marked as `ToBeRemoved`
- Adding Rebar object to the Physical_oM

### CODE REVIEW

**Comment by pawelbaran** [comment]
@LMarkowski we had a chat with @IsakNaslundBh about different classes vs enum defining function - ended up with classes due to the fact that at this stage we are likely not to predict need for extra properties specific to certain subtypes of rebar. Looks like you have just proven our assumption, suggesting that stirrups will need an extra property defining hooks.

However, would be good to hear opinions of more experienced ones, @rwemay @JosefTaylor @peterjamesnugent?

File: Structure_oM/SectionProperties/Reinforcement/LayerReinforcement.cs [hunk 1]
```diff
@@ -20,8 +20,11 @@
  * along with this code. If not, see <https://www.gnu.org/licenses/lgpl-3.0.html>.      
  */
 
+using BH.oM.Reflection.Attributes;
+
 namespace BH.oM.Structure.SectionProperties.Reinforcement
 {
+    [ToBeRemoved("3.2", "Complete refactoring of reinforcement is being done, outside current possible scope of versioning. Old reinfrcement classes will be removed.")]
```
Comment (FraserGreenroyd) [inline Structure_oM/SectionProperties/Reinforcement/LayerReinforcement.cs:27]: Sp: `reinforcement`

File: Structure_oM/SectionProperties/Reinforcement/TieReinforcement.cs [hunk 2]
```diff
@@ -19,9 +19,11 @@
  * You should have received a copy of the GNU Lesser General Public License     
  * along with this code. If not, see <https://www.gnu.org/licenses/lgpl-3.0.html>.      
  */
+using BH.oM.Reflection.Attributes;
 
 namespace BH.oM.Structure.SectionProperties.Reinforcement
 {
+    [ToBeRemoved("3.2", "Complete refactoring of reinforcement is being done, outside current possible scope of versioning. Old reinfrcement classes will be removed.")]
```
Comment (FraserGreenroyd) [inline Structure_oM/SectionProperties/Reinforcement/TieReinforcement.cs:26]: Sp. `reinforcement`

File: Structure_oM/SectionProperties/Reinforcement/PerimeterReinforcement.cs [hunk 3]
```diff
@@ -19,12 +19,11 @@
...
  * along with this code. If not, see <https://www.gnu.org/licenses/lgpl-3.0.html>.      
  */
+using BH.oM.Reflection.Attributes;
 
 namespace BH.oM.Structure.SectionProperties.Reinforcement
 {
-    /// <summary>
-    /// Perimeter Reinforcement is aimed at columns and is only valid on rectangular and circular sections
-    /// </summary>
+    [ToBeRemoved("3.2", "Complete refactoring of reinforcement is being done, outside current possible scope of versioning. Old reinfrcement classes will be removed.")]
```
Comment (FraserGreenroyd) [inline Structure_oM/SectionProperties/Reinforcement/PerimeterReinforcement.cs:26]: Sp. `reinforcement`

File: Structure_oM/SectionProperties/Reinforcement/Reinforcement.cs [hunk 4]
```diff
@@ -22,9 +22,11 @@
 
 using BH.oM.Base;
 using BH.oM.Physical.Materials;
+using BH.oM.Reflection.Attributes;
 
 namespace BH.oM.Structure.SectionProperties.Reinforcement
 {
+    [ToBeRemoved("3.2", "Complete refactoring of reinforcement is being done, outside current possible scope of versioning. Old reinfrcement classes will be removed.")]
```
Comment (FraserGreenroyd) [inline Structure_oM/SectionProperties/Reinforcement/Reinforcement.cs:29]: Sp. `reinforcement`

**Comment by JosefTaylor** [comment]
> However, would be good to hear opinions of more experienced ones, @rwemay @JosefTaylor @peterjamesnugent?

I think we will end up with two types of bars: one which is defined by a point and extents, i.e. a longitudinal bar, or a wall face bar, etc.; and one which is defined by a sketch, i.e. stirrups, hoops, hairpins, bent bars etc.. This lines up with rebar fabricator's formats, where they call out straight bars as (1)#5 x 8' long, and have a bend table for all the bent bars and stirrups.

I think an enum parameter for development type could be useful for longitudinal bars - will it be developed straight, 90 hook, 180 hook, terminal, etc.. That could get us from design layouts to cut layouts. I don't think we need to develop that functionality right now, but it would be useful in the future.

**Comment by IsakNaslundBh** [comment]
Updating from master to get CI to behave. @LMarkowski do you mind re-checking?

File: Structure_oM/SectionProperties/Reinforcement/LongitudinalReinforcement.cs [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+
+using BH.oM.Base;
+using BH.oM.Structure.MaterialFragments;
+using BH.oM.Spatial.Layouts;
+using BH.oM.Quantities.Attributes;
+using System.ComponentModel;
+
+namespace BH.oM.Structure.SectionProperties.Reinforcement
+{
+    [Description("Defines any longitudinal reinforcement along a Bar or IFramingElement.")]
```
Comment (kThorsager) [inline Structure_oM/SectionProperties/Reinforcement/LongitudinalReinforcement.cs:31]:
```suggestion
    [Description("Defines any longitudinal reinforcement along a Bar.")]
```

File: Structure_oM/SectionProperties/Reinforcement/LongitudinalReinforcement.cs [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+    public class LongitudinalReinforcement : BHoMObject, IBarReinforcement
+    {
+        /***************************************************/
+        /**** Properties                                ****/
+        /***************************************************/
+
+        [Description("Layout controling the distribution of the rebars across the SectionProperty")]
```
Comment (kThorsager) [inline Structure_oM/SectionProperties/Reinforcement/LongitudinalReinforcement.cs:38]:
```suggestion
        [Description("Layout controlling the distribution of the rebars across the SectionProperty.")]
```

**Comment by IsakNaslundBh** [comment]
Thanks @kThorsager . Have applied your suggestions

**Pull request merged**